src/ToricVarieties/ToricMorphisms/attributes.jl

```diff
@@ -0,0 +1,199 @@
+
+julia> codomain(ToricIdentityMorphism(F4))
+A normal, non-affine, smooth, projective, gorenstein, non-fano, 2-dimensional toric variety without torusfactor
+```
+"""
+codomain(tm::ToricMorphism) = tm.codomain
+export codomain
+
+
+@doc Markdown.doc"""
+    image(tm::ToricMorphism)
+
+Return the image of the toric morphism `tm`.
+
+# Examples
+```jldoctest
+julia> F4 = hirzebruch_surface(4)
+A normal, non-affine, smooth, projective, gorenstein, non-fano, 2-dimensional toric variety without torusfactor
+
+julia> image(ToricIdentityMorphism(F4))
+A normal, non-affine, smooth, projective, gorenstein, non-fano, 2-dimensional toric variety without torusfactor
+```
+"""
+image(tm::ToricMorphism) = tm.image
+export image
+
+
+@doc Markdown.doc"""
+    grid_morphism(tm::ToricMorphism)
+
+Return the underlying grid morphism of the toric morphism `tm`.
+
+# Examples
+```jldoctest
+julia> F4 = hirzebruch_surface(4)
+A normal, non-affine, smooth, projective, gorenstein, non-fano, 2-dimensional toric variety without torusfactor
+
+julia> grid_morphism(ToricIdentityMorphism(F4))
+Map with following data
+Domain:
+=======
+Abelian group with structure: Z^2
+Codomain:
+=========
+Abelian group with structure: Z^2
+```
+"""
+grid_morphism(tm::ToricMorphism) = tm.grid_morphism
+export grid_morphism
+
+
+
+@doc Markdown.doc"""
+    morphism_on_torusinvariant_weil_divisor_group(tm::ToricMorphism)
+
+For a given toric morphism `tm`, this method computes the corresponding
+map of the torusinvariant Weil divisors.
+
+# Examples
+```jldoctest
+julia> F4 = hirzebruch_surface(4)
+A normal, non-affine, smooth, projective, gorenstein, non-fano, 2-dimensional toric variety without torusfactor
+
+julia> morphism_on_torusinvariant_weil_divisor_group(ToricIdentityMorphism(F4))
+Map with following data
+Domain:
+=======
+Abelian group with structure: Z^4
+Codomain:
+=========
+Abelian group with structure: Z^4
+```
+"""
+@attr GrpAbFinGenMap function morphism_on_torusinvariant_weil_divisor_group(tm::ToricMorphism)
+
+    # find domain and codomain variety
+    domain_variety = domain(tm)
+    codomain_variety = codomain(tm)
+
+    # find rays of these varieties
+    domain_variety_rays = matrix(ZZ,rays(domain_variety))
+    codomain_variety_rays = matrix(ZZ,rays(codomain_variety))
+
+    # identify the image rays
+    images_of_rays = domain_variety_rays * matrix(grid_morphism(tm))
+
+    # for each image ray, find a max cone in the codomain toric variety which contains it
+    codomain_variety_max_cones = maximal_cones(codomain_variety)
+    rayimage_maxcones_incidence_list = [0 for i in 1:nrows(images_of_rays)]
+    for i in 1:nrows(images_of_rays)
+      for j in 1:length(codomain_variety_max_cones)
+        if contains(codomain_variety_max_cones[j], [images_of_rays[i,k] for k in 1:ncols(images_of_rays)])
+          rayimage_maxcones_incidence_list[i] = j
+          break
+        end
+      end
+    end
```

Comment:
This is duplicating a lot of checks, isn't it? Also I believe Max said that there is a way to avoid nested for loops, I think you can just separate the two params with a comma.

Reply:
Nevermind, I was just confused. Can you not write something like `images_of_rays[i,:]`?

Reply:
Also what if a ray is contained in several maximal cones?